pkg/plumbing/watch/server.go

```diff
@@ -0,0 +1,290 @@
+const (
+	LogTag_Server         = "╬═  server"
+	LogTag_DefaultHandler = "╬═? handler" // Default log tag for handler that doesn't have a connection ID
+)
+
+// key for context.Context used to set/retrieve handler logging tag
+type handlerTagKey struct{}
+
+// handlerTag retrieves the handler logging tag from context
+func handlerTag(ctx context.Context) string {
+	value := ctx.Value(handlerTagKey{})
+	if value == nil {
+		return LogTag_DefaultHandler
+	}
+	return value.(string)
+}
+
+// setHandlerTag returns a new context with the given handler logging tag value
+func setHandlerTag(ctx context.Context, tag string) context.Context {
+	return context.WithValue(ctx, handlerTagKey{}, tag)
+}
+
+// server stores the current status of the plot execution and responds to clients.
+type server struct {
+	listener    net.Listener     // Where connections come from.
+	handler     rpcHandler       // Handles actual RPCs.
+	nowFn       func() time.Time // Should return _now_ for the purposes of setting deadlines. If nil, time.Now will be used.
+	readTimeout *time.Duration   // Timeout for deadlines.
+	acceptCount int              // Tracing metric. Used in log tags for handlers.
+}
+
+const DefaultReadTimeout = 5 * time.Second // Default read timeout for connections. See net.Conn.SetReadDeadline, server.setReadDeadline.
+
+// now returns the result of time.Now
+// May be overridden by setting nowFn
+func (s *server) now() time.Time {
+	if s.nowFn == nil {
+		return time.Now()
+	}
+	return s.nowFn()
+}
+
+// setReadDeadline sets the connection read deadline based on server configuration.
+// This should prevent connections from getting stuck waiting for a client request.
+// The default deadline is DefaultReadTimout from time.Now.
+//
+// Errors:
+//
+//   - warpforge-error-internal -- failed to set connection read deadline
+func (s *server) setReadDeadline(conn net.Conn) error {
+	to := DefaultReadTimeout
+	if s.readTimeout != nil {
+		to = *s.readTimeout
+	}
+	deadline := s.now().Add(to)
+	err := conn.SetReadDeadline(deadline)
+	// err := conn.SetDeadline(deadline)
+	if err != nil {
+		return serum.Error(wfapi.ECodeInternal,
+			serum.WithCause(err),
+			serum.WithMessageTemplate("server failed to set read deadline {{deadline}}"),
+			serum.WithDetail("deadline", deadline.String()),
+		)
+	}
+	return nil
+}
+
+// NextRPC returns the next RPC object on the decoder's stream.
+//
+// Errors:
+//
+//    - warpforge-error-rpc-serialization -- bad connection or invalid json
+//    - warpforge-error-rpc-serialization -- invalid RPC data
+func NextRPC(ctx context.Context, d *json.Decoder) (*workspaceapi.Rpc, error) {
+	var err error
+	var raw json.RawMessage
+	var rpc workspaceapi.Rpc
+	log := logging.Ctx(ctx)
+	tag := handlerTag(ctx)
+	log.Debug(tag, "reading json")
+	err = d.Decode(&raw)
+	if err != nil {
+		return nil, serum.Error(workspaceapi.ECodeRpcSerialization, serum.WithCause(err),
```

Comment:
This error gets hit twice after every valid RPC. I believe we're hitting an EOF on `NextRPC` then another EOF when the connection is closed. Here is the full log for a single RPC request:

```
╬═  server[26]  getting next RPC
╬═  server[26]  handle request
╬═  server[26]  response
╬═  server[26]  getting next RPC
╬═  server[26]  unable to read RPC: warpforge-error-rpc-serialization: unable to read JSON from connection: caused by: bestguess-golang-errors-errorString: EOF
╬═  server[26]  connection closed
╬═  server[26]  handler returned with error: warpforge-error-rpc-serialization: unable to read JSON from connection: caused by: bestguess-golang-errors-errorString: EOF
```

Reply:
Yeah I've been ignoring that... I suppose I should handle it more gracefully.

Reply:
Server output should be debug only and it should ignore the EOF caused by the client dropping connections.